### PR TITLE
Rector itself - merge isCandidate() to refactor() + add getNodeType()

### LINKED ISSUE #553: Consider changing isCandidate() to Node class subsriber
Since PHP_CodeSniffer, PHP CS Fixer and PHPStan use the same approach: https://github.com/spaze/phpstan-disallowed-calls/blob/master/src/MethodCalls.php

And its not often clear where to put condition and "active..." helper properties are used.

Early returns might work better and also faster, due to only token subscription.

### PR DESCRIPTION
**Pull request by TomasVotruba**
Closes #553

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -45,7 +45,8 @@
             "Rector\\Doctrine\\": "packages/Doctrine/src",
             "Rector\\Utils\\": "packages/Utils/src",
             "Rector\\YamlRector\\": "packages/YamlRector/src"
-        }
+        },
+        "classmap": ["custom"]
```
Comment (carusogabriel) [inline composer.json]: Why?

File: custom/MergeIsCandidateRector.php [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+use Rector\Builder\StatementGlue;
+use Rector\NodeTypeResolver\NodeTypeResolver;
+use Rector\NodeTypeResolver\PhpDoc\NodeAnalyzer\DocBlockAnalyzer;
+use Rector\NodeTypeResolver\PHPStan\Type\TypeToStringResolver;
+use Rector\Rector\AbstractRector;
+use Rector\RectorDefinition\CodeSample;
+use Rector\RectorDefinition\RectorDefinition;
+use Rector\Utils\NodeTraverser\CallableNodeTraverser;
+
+final class MergeIsCandidateRector extends AbstractRector
```
Comment (carusogabriel) [inline custom/MergeIsCandidateRector.php]: Will we drop it before merge this PR?

File: packages/PHPUnit/src/Rector/ExceptionAnnotationRector.php [hunk 3]
```diff
@@ -73,30 +73,23 @@ public function test()
...
-        return false;
+        return [ClassMethod::class];
     }
 
     /**
      * @param ClassMethod $classMethodNode
      */
     public function refactor(Node $classMethodNode): ?Node
     {
+        if (! $classMethodNode instanceof ClassMethod) {
```
Comment (carusogabriel) [inline packages/PHPUnit/src/Rector/ExceptionAnnotationRector.php]: I don't think we need it. If the refactor method was called, it means we have an instance of `ClassMethod`, like returned in `getNodeTypes`. This should be fixed in all Rectors

File: src/Rector/AbstractRector.php [hunk 4]
```diff
@@ -50,17 +50,24 @@ public function setAbstractRectorDependencies(PropertyAdder $propertyAdder, Expr
...
+//        } elseif (method_exists($this, 'isCandidate')) {
+//            if (! $this->isCandidate($node)) {
+//                return null;
+//            }
+//        }
 
         $newNode = $this->refactor($node);
+
+        // nothing has changed, return original node
         if ($newNode !== null) {
```
Comment (carusogabriel) [inline src/Rector/AbstractRector.php]: Wrong comment? If `$newNode` is different from null, it means we have a new Node

File: src/Rector/AbstractRector.php [hunk 5]
```diff
@@ -50,17 +50,24 @@ public function setAbstractRectorDependencies(PropertyAdder $propertyAdder, Expr
...
-        } elseif (method_exists($this, 'isCandidate')) {
-            if (! $this->isCandidate($node)) {
-                return null;
-            }
+        $nodeClass = get_class($node);
+
+        if (! in_array($nodeClass, $this->getNodeTypes(), true)) {
+            return null;
         }
+//        if (method_exists($this, 'getNodeTypes')) {
```
Comment (carusogabriel) [inline src/Rector/AbstractRector.php]: TODO: Remove it and add `getNodeTypes` as an abstract method

File: composer.json (hunk 1 above)
Comment (TomasVotruba) [reply composer.json]: So composer can load `custom/MergeIsCandidateRector.php`.

I'll move it to `/examples` in the end, so we have complex code to show people.

File: custom/MergeIsCandidateRector.php (hunk 2 above)
Comment (TomasVotruba) [reply custom/MergeIsCandidateRector.php]: Yea, I'll just move it to `/examples`, to show what Rector can do.

File: packages/PHPUnit/src/Rector/ExceptionAnnotationRector.php (hunk 3 above)
Comment (TomasVotruba) [reply packages/PHPUnit/src/Rector/ExceptionAnnotationRector.php]: Yes yes, working on it!

**Comment by TomasVotruba** [comment]
- duplicated old `instanceof` removed
- rector `MergeIsCandidateRector` moved to examples
- added `Rector\Utils\NodeTraverser\CallableNodeTraverser` that helps with inner traversing
- rebased

Ready to merge :+1:

**Comment by carusogabriel** [comment]
@TomasVotruba I'll review this afternoon and I'll merge if everything is all right :+1:

**Comment by TomasVotruba** [comment]
:+1: Thanks @carusogabriel

**Pull request merged**